extensions-core/spark-extensions/src/main/scala/org/apache/druid/spark/clients/DruidClient.scala

```diff
@@ -0,0 +1,180 @@
+
+import java.net.URL
+import java.util.{List => JList}
+import javax.ws.rs.core.MediaType
+import scala.collection.JavaConverters.{asScalaBufferConverter, mapAsJavaMapConverter,
+  mapAsScalaMapConverter, seqAsJavaListConverter}
+
+/**
+  * Going with SegmentMetadataQueries despite the significant overhead because there's no other way
+  * to get accurate information about Druid columns.
+  *
+  * This is pulled pretty directly from the druid-hadoop-input-format project on GitHub. There is
+  * likely substantial room for improvement.
+  *
+  * @param httpClient
+  * @param objectMapper
+  * @param hostAndPort
+  */
+class DruidClient(
+                   val httpClient: HttpClient,
+                   val objectMapper: ObjectMapper,
+                   val hostAndPort: HostAndPort
+                 ) extends Logging {
+  private val RETRY_COUNT = 5
+  private val RETRY_WAIT_SECONDS = 5
+  private val TIMEOUT_MILLISECONDS = 300000
+  private val druidBaseQueryURL: HostAndPort => String =
+    (hostAndPort: HostAndPort) => s"http://$hostAndPort/druid/v2/"
+
+  /**
+    * The SQL system catalog tables are incorrect for multivalue columns and don't have accurate
+    * type info, so we need to fall back to segmentMetadataQueries. Given a DATASOURCE and a range
+    * of INTERVALS to query over, return a map from column name to a tuple of
+    * (columnType, hasMultipleValues). Note that this is a very expensive operation over large
+    * numbers of segments. If possible, this method should only be called with the most granular
+    * starting and ending intervals instead of over a larger interval.
+    *
+    * @param dataSource The Druid dataSource to fetch the schema for.
+    * @param intervals  The intervals to return the schema for.
+    * @return A map from column name to data type and whether or not the column is multi-value
+    *         for the schema of DATASOURCE.
+    */
+  def getSchema(dataSource: String, intervals: List[Interval]): Map[String, (String, Boolean)] = {
+    val body = Druids.newSegmentMetadataQueryBuilder()
+      .dataSource(dataSource)
+      .intervals(intervals.asJava)
+      .analysisTypes(SegmentMetadataQuery.AnalysisType.SIZE)
+      .merge(true)
+      .context(Map[String, AnyRef](
+        "timeout" -> Int.box(TIMEOUT_MILLISECONDS)
+      ).asJava)
+      .build()
+    val response = sendRequestWithRetry(
+      druidBaseQueryURL(hostAndPort), RETRY_COUNT, Option(objectMapper.writeValueAsBytes(body))
+    )
+    val segments =
+      objectMapper.readValue[JList[SegmentAnalysis]](
+        response.getContent, new TypeReference[JList[SegmentAnalysis]]() {})
+    if (segments.size() == 0) {
+      throw new ISE(
+        s"No segments found for intervals [${intervals.mkString(",")}] on $dataSource"
+      )
+    }
+    // Since we're setting merge to true, there should only be one item in the list
+    if (segments.size() > 1) {
+      logWarn("Segment metadata response had more than one row, problem?")
```

Comment:
I am not sure if simply logging a WARN message is enough. I would rather fail the request since this shouldn't happen.

Reply:
Yeah, this should error. Good catch.